# Data Exploration Notebook

This notebook is for prototyping and exploring the data before moving code into the application files.

## Step 1: Import Libraries

In [1]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
from pathlib import Path

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

print("Libraries imported successfully!")

Libraries imported successfully!


In [2]:
# Install required packages for Jupyter notebooks
import subprocess
import sys

# Check and install nbformat
try:
    import nbformat
    print("nbformat already installed!")
except ImportError:
    print("Installing nbformat...")
    subprocess.check_call([sys.executable, "-m", "pip", "install", "nbformat", "-q"])
    print("nbformat installed successfully!")

# Check and install openpyxl (needed for Excel files)
try:
    import openpyxl
    print("openpyxl already installed!")
except ImportError:
    print("Installing openpyxl...")
    subprocess.check_call([sys.executable, "-m", "pip", "install", "openpyxl", "-q"])
    print("openpyxl installed successfully!")

print("Setup complete!")

nbformat already installed!
openpyxl already installed!
Setup complete!


## Step 2: Load Grossi Data

Focus on the Grossi dataset with multiple sheets

In [3]:
# Define data paths
data_dir = Path('../data')

# Load Grossi Excel file (contains multiple sheets)
grossi_excel = data_dir / 'Grossi_et_al._Piloted_data_extraction_strategy.xlsx'

print(f"Data directory: {data_dir}")
print(f"Grossi Excel file exists: {grossi_excel.exists()}")

Data directory: ..\data
Grossi Excel file exists: True


## Step 3: Explore All Sheets in Grossi Excel

In [4]:
# Check if Excel file has multiple sheets
grossi_excel = data_dir / 'Grossi_et_al._Piloted_data_extraction_strategy.xlsx'

if grossi_excel.exists():
    xls = pd.ExcelFile(grossi_excel)
    print(f"Number of sheets: {len(xls.sheet_names)}")
    print(f"\nSheet names:")
    for i, sheet in enumerate(xls.sheet_names, 1):
        print(f"  {i}. {sheet}")
else:
    print("Excel file not found, using CSV only")

Number of sheets: 5

Sheet names:
  1. README
  2. Bibliography
  3. Data extraction
  4. Typologies
  5. Extended definitions


In [5]:
# Load all sheets from Excel file (if it exists)
if grossi_excel.exists():
    sheets_data = {}
    for sheet_name in xls.sheet_names:
        df_sheet = pd.read_excel(grossi_excel, sheet_name=sheet_name)
        sheets_data[sheet_name] = df_sheet
        print(f"\n{sheet_name}: {df_sheet.shape[0]} rows, {df_sheet.shape[1]} columns")
        print(f"  Columns: {df_sheet.columns.tolist()[:5]}...")  # First 5 columns
    
    print(f"\n✓ Loaded {len(sheets_data)} sheets into 'sheets_data' dictionary")
else:
    print("Using CSV data only")


README: 1 rows, 1 columns
  Columns: ['README']...

Bibliography: 16 rows, 13 columns
  Columns: ['1.1', '1.2', '1.3', '1.4', '1.5']...

Data extraction: 16 rows, 27 columns
  Columns: [1.1, '1.2', '2.1', '2.2', '2.3']...

Typologies: 261 rows, 25 columns
  Columns: ['Typologies', 'Unnamed: 1', '*Where multiple, use a semi-colon to delimit (applies to: 3.1; 3.3; 4.3; 6.1; 6.2; 6.3; 7.1; 7.2)', 'Unnamed: 3', 'Unnamed: 4']...

Extended definitions: 76 rows, 5 columns
  Columns: ['Extended Definitions', 'Unnamed: 1', 'Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4']...

✓ Loaded 5 sheets into 'sheets_data' dictionary


### Understanding Each Sheet

Based on the output:
1. **README** - 1 row (metadata/description)
2. **Bibliography** - 16 rows, 13 columns (keyword definitions/references)
3. **Data extraction** - 16 rows, 27 columns (main data)
4. **Typologies** - 261 rows, 25 columns (classification/categories)
5. **Extended definitions** - 76 rows, 5 columns (detailed definitions)

Let's explore each one:

In [6]:
# 1. Data extraction sheet - Main data
df_data_extraction = sheets_data['Data extraction']
print("=== DATA EXTRACTION SHEET ===")
print(f"Shape: {df_data_extraction.shape}")
print(f"\nColumns: {df_data_extraction.columns.tolist()}")
print("\nFirst row:")
df_data_extraction.head(1)

=== DATA EXTRACTION SHEET ===
Shape: (16, 27)

Columns: [1.1, '1.2', '2.1', '2.2', '2.3', '2.4', '2.5', '3.1', 3.2, 3.3, '3.4', '3.5', '4.1', '4.2', '4.3', '4.4', '4.5', '4.6', '5.1', '5.2', '5.3', '6.1', '6.2', '6.3', '6.4', '6.5', '7.1']

First row:


,1.1,1.2,2.1,2.2,2.3,2.4,2.5,3.1,3.2,3.3,3.4,3.5,4.1,4.2,4.3,4.4,4.5,4.6,5.1,5.2,5.3,6.1,6.2,6.3,6.4,6.5,7.1
0,Article_ID,Authors,Date_Coded,Coded_by,Inclusion,IfExcluded_Criteria,IfExcluded_Comment,Study_design,Ecoregion,Spatial_Scale,Continent_Ocean,Country,Threat_metric,Threat_data,Threat,Quantity_threats,Threat_precision,Threat_database,Plant_group,Animal_group,Ecological_level,Direct_driver,Indirect_driver,Georef_ind_driver,Ind_driver_text_source,Spatial_scale_text_source,Biodiversity_impacts


In [7]:
# 2. Bibliography sheet - Keyword meanings
df_bibliography = sheets_data['Bibliography']
print("=== BIBLIOGRAPHY SHEET ===")
print(f"Shape: {df_bibliography.shape}")
print(f"\nColumns: {df_bibliography.columns.tolist()}")
print("\nSample data:")
df_bibliography.head(3)

=== BIBLIOGRAPHY SHEET ===
Shape: (16, 13)

Columns: ['1.1', '1.2', '1.3', '1.4', '1.5', '1.6', '1.7', '1.8', '1.9', '1.10', '1.11', '1.12', '1.13']

Sample data:


,1.1,1.2,1.3,1.4,1.5,1.6,1.7,1.8,1.9,1.10,1.11,1.12,1.13
0,Article_ID,Date_screened,Screened_by,Pub_Type,Authors,Year,Title,Pub_name,Volume,Issue,Article_Type,Pages,URL
1,93,26/07/2024,1,Journal Article,ABRAM_etal,2020,Mapping perceptions of species' threats and po...,Diversity and Distributions,21,5,Article,487-499,https://www.scopus.com/inward/record.uri?eid=2...
2,108,26/07/2024,1,Journal Article,BIRKMANIS_etal,2009,Shark conservation hindered by lack of habitat...,Global Ecology and Conservation,21,NaN,Article,NaN,https://www.scopus.com/inward/record.uri?eid=2...


In [8]:
# 3. Typologies sheet - Classifications/categories
df_typologies = sheets_data['Typologies']
print("=== TYPOLOGIES SHEET ===")
print(f"Shape: {df_typologies.shape}")
print(f"\nColumns: {df_typologies.columns.tolist()}")
print("\nSample data:")
df_typologies.head(5)

=== TYPOLOGIES SHEET ===
Shape: (261, 25)

Columns: ['Typologies', 'Unnamed: 1', '*Where multiple, use a semi-colon to delimit (applies to: 3.1; 3.3; 4.3; 6.1; 6.2; 6.3; 7.1; 7.2)', 'Unnamed: 3', 'Unnamed: 4', 'Unnamed: 5', 'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9', 'Unnamed: 10', 'Unnamed: 11', 'Unnamed: 12', 'Unnamed: 13', 'Unnamed: 14', 'Unnamed: 15', 'Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18', 'Unnamed: 19', 'Unnamed: 20', 'Unnamed: 21', 'Unnamed: 22', 'Unnamed: 23', 'Unnamed: 24']

Sample data:


,Typologies,Unnamed: 1,"*Where multiple, use a semi-colon to delimit (applies to: 3.1; 3.3; 4.3; 6.1; 6.2; 6.3; 7.1; 7.2)",Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6,Unnamed: 7,Unnamed: 8,Unnamed: 9,Unnamed: 10,Unnamed: 11,Unnamed: 12,Unnamed: 13,Unnamed: 14,Unnamed: 15,Unnamed: 16,Unnamed: 17,Unnamed: 18,Unnamed: 19,Unnamed: 20,Unnamed: 21,Unnamed: 22,Unnamed: 23,Unnamed: 24
0,Description,NaN,The coding table is sourced from Ridley et al....,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Field Code,1.1,2.1,2.2,2.3,3.1,3.2,3.3,3.4,3.5,4.1,4.2,4.3,4.4,4.5,4.6,5.1,5.2,5.3,6.1,6.2,6.3,6.4,6.5,7.1
2,Data Field,Article_ID,Date_Coded,Coded_by,Inclusion,Study_design,Ecoregion,Spatial_scale,Continent_Ocean,Country_EEZ,Threat_metric,Threat_data,Threat,Quantity_threats,threat_precision,Threat_database,Plant_taxon_group,Animal_taxon_group,Ecological_level,Direct_driver,Indirect_driver,Georef_ind_driver,Ind_driver_text_source,Spatial_scale_text_source,Biodiversity_impacts
3,Data Type,text,date,text,binary,text,text,text,text,text,text,text,list,numeric,list,text,text,text,list,list,list,binary,text,text,list
4,Info,NaN,Removed after coding is consolidated between t...,NaN,NaN,NaN,NaN,NaN,na applied if global,na applied if spatial scale > country,NaN,NaN,NaN,Number of tier 2 threats (3.6),NaN,name of data source,class,class,NaN,NaN,na applied if indirect driver is not mentioned...,na applied if spatial scale is not mentioned i...,specific sentence where the indirect driver is...,specific sentence where the spatial scale of t...,NaN


In [9]:
# 4. Extended definitions sheet - Detailed definitions
df_extended = sheets_data['Extended definitions']
print("=== EXTENDED DEFINITIONS SHEET ===")
print(f"Shape: {df_extended.shape}")
print(f"\nColumns: {df_extended.columns.tolist()}")
print("\nSample data:")
df_extended.head(5)

=== EXTENDED DEFINITIONS SHEET ===
Shape: (76, 5)

Columns: ['Extended Definitions', 'Unnamed: 1', 'Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4']

Sample data:


,Extended Definitions,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4
0,NaN,NaN,NaN,NaN,NaN
1,3.1 Study Design,NaN,NaN,NaN,NaN
2,NaN,Observational,NaN,NaN,Variables are not actively manipulated
3,NaN,Experimental,NaN,NaN,For the sake of this analysis all research whi...
4,NaN,S_Review,NaN,NaN,Systematic review (including meta-analyses). N...


### Sheet Relationships

Based on our understanding:
- **Data extraction** = Main dataset for analysis
- **Bibliography** = Reference/definitions for keywords in data extraction
- **Typologies** = Likely contains classification schemes/categories used in data extraction
- **Extended definitions** = More detailed explanations of terms/concepts

Let's focus on the Data extraction sheet and understand its structure better:

In [10]:
# Clean the Data extraction sheet - skip the first row with numbers
df_main = pd.read_excel(grossi_excel, sheet_name='Data extraction', skiprows=1)
print("=== CLEANED DATA EXTRACTION ===")
print(f"Shape: {df_main.shape}")
print(f"\nColumn names:")
for i, col in enumerate(df_main.columns, 1):
    print(f"  {i}. {col}")
    
print("\n✓ This is the main dataset we'll use for analysis")

=== CLEANED DATA EXTRACTION ===
Shape: (15, 27)

Column names:
  1. Article_ID
  2. Authors
  3. Date_Coded
  4. Coded_by
  5. Inclusion
  6. IfExcluded_Criteria
  7. IfExcluded_Comment
  8. Study_design
  9. Ecoregion
  10. Spatial_Scale 
  11. Continent_Ocean
  12. Country
  13. Threat_metric
  14. Threat_data
  15. Threat
  16. Quantity_threats
  17. Threat_precision
  18. Threat_database
  19. Plant_group
  20. Animal_group
  21. Ecological_level
  22. Direct_driver
  23. Indirect_driver
  24. Georef_ind_driver
  25. Ind_driver_text_source
  26. Spatial_scale_text_source
  27. Biodiversity_impacts

✓ This is the main dataset we'll use for analysis


In [11]:
# Preview the clean data
df_main.head()

,Article_ID,Authors,Date_Coded,Coded_by,Inclusion,IfExcluded_Criteria,IfExcluded_Comment,Study_design,Ecoregion,Spatial_Scale,Continent_Ocean,Country,Threat_metric,Threat_data,Threat,Quantity_threats,Threat_precision,Threat_database,Plant_group,Animal_group,Ecological_level,Direct_driver,Indirect_driver,Georef_ind_driver,Ind_driver_text_source,Spatial_scale_text_source,Biodiversity_impacts
0,93,Abram_etal,16/08/2024,FG,1,NaN,NaN,Observational,Terrestrial,Local,asia,Indonesia,expert,Quantitative,6.3:Other,1,level2,na,na,mammal,population,land/sea_use_change,economic;society_and_culture;governance_and_in...,1,Higher rates of killings in the region may be ...,Higher rates of killings in the region may be ...,decrease
1,108,Birkmanis_etal,16/08/2024,FG,1,NaN,NaN,Observational,Marine,National,indian ocean,Australia,database,Quantitative,5.1:UseAnimals,1,level2,AustalianBureauofAgricultural and ResourceEcon...,na,fish,population,direct_exploitation,governance_and_institutions,1,"Using Australia as a case study, we found that...","Using Australia as a case study, we found that...",decrease
2,61,Chu_etal,16/08/2024,FG,1,NaN,NaN,Observational,Freshwater,National,north america,Canada,database,Quantitative,2.1:AgNTC;2.2:AgWood;9.1:PollDWW;9.4:PollGarbS...,6,cumulative,AgricultureAndAgri-FoodCanada_1997,na,fish,biome,land/sea_use_change;pollution,economic;demographic,1,"Our study identified dwelling, road density, c...","Our study identified dwelling, road density, c...",na
3,90,Desta and Fetene,16/08/2024,FG,1,NaN,NaN,Observational,Freshwater;Terrestrial,Local,africa,Ethiopia,rs,Qualitative,1.1:DevResid;2.1:AgNTC;2.2:AgWood;2.3:AgLivestock,4,level2,na,forest,na,ecosystem,land/sea_use_change,economic;demographic;governance_and_institutions,1;na,Ethiopia’s Lake Ziway watershed supports the l...,Ethiopia’s Lake Ziway watershed supports the l...,decrease
4,112,Di Fulvio_etal,25/07/2024,FG,1,NaN,NaN,Observational,Terrestrial,Continental,europe,na,modelled,Quantitative,2.1:AgNTC;3.3:Renew,1,level2,na,tracheophyta,mammal;amphibia;bird;reptile,biome,land/sea_use_change,economic,0,Our results show that policies promoting bioen...,Our results show that policies promoting bioen...,na


## Step 4: Work with Main Data Extraction Sheet

Now that we have the clean data, let's start analyzing it

In [12]:
# Explore the main data columns and values
print("=== MAIN DATA OVERVIEW ===")
print(f"Shape: {df_main.shape}")
print(f"\nColumn names:")
for i, col in enumerate(df_main.columns, 1):
    print(f"  {i}. {col}")

# Check for key columns
print("\n=== KEY COLUMNS ===")
if 'Continent_Ocean' in df_main.columns:
    print(f"Unique Continents/Oceans: {df_main['Continent_Ocean'].nunique()}")
    print(df_main['Continent_Ocean'].value_counts())
else:
    print("Continent_Ocean column not found")

=== MAIN DATA OVERVIEW ===
Shape: (15, 27)

Column names:
  1. Article_ID
  2. Authors
  3. Date_Coded
  4. Coded_by
  5. Inclusion
  6. IfExcluded_Criteria
  7. IfExcluded_Comment
  8. Study_design
  9. Ecoregion
  10. Spatial_Scale 
  11. Continent_Ocean
  12. Country
  13. Threat_metric
  14. Threat_data
  15. Threat
  16. Quantity_threats
  17. Threat_precision
  18. Threat_database
  19. Plant_group
  20. Animal_group
  21. Ecological_level
  22. Direct_driver
  23. Indirect_driver
  24. Georef_ind_driver
  25. Ind_driver_text_source
  26. Spatial_scale_text_source
  27. Biodiversity_impacts

=== KEY COLUMNS ===
Unique Continents/Oceans: 9
Continent_Ocean
na               4
asia             2
africa           2
mediterranean    2
indian ocean     1
north america    1
europe           1
south america    1
north pacific    1
Name: count, dtype: int64


In [17]:
# Create visualizations with the main data
from IPython.display import display

if 'Continent_Ocean' in df_main.columns:
    continent_counts = df_main['Continent_Ocean'].value_counts()
    
    fig = px.bar(
        x=continent_counts.index,
        y=continent_counts.values,
        labels={'x': 'Continent/Ocean', 'y': 'Count'},
        title='Distribution by Continent/Ocean - Grossi Data'
    )
    display(fig)

else:    print("Cannot create chart - Continent_Ocean column not found")

In [18]:
# Explore threat types
if 'Threat' in df_main.columns:
    threat_counts = df_main['Threat'].value_counts().head(10)
    
    fig = px.bar(
        x=threat_counts.values,
        y=threat_counts.index,
        orientation='h',
        labels={'x': 'Count', 'y': 'Threat Type'},
        title='Top 10 Threat Types'
    )
    display(fig)
else:
    print("Threat column not found")

In [20]:
# Explore study design
if 'Study_design' in df_main.columns:
    design_counts = df_main['Study_design'].value_counts()
    
    fig = px.pie(
        values=design_counts.values,
        names=design_counts.index,
        title='Study Design Distribution'
    )
    display(fig)
else:
    print("Study_design column not found")

In [21]:
# Explore ecological levels
if 'Ecological_level' in df_main.columns:
    eco_counts = df_main['Ecological_level'].value_counts()
    
    fig = px.bar(
        x=eco_counts.index,
        y=eco_counts.values,
        labels={'x': 'Ecological Level', 'y': 'Count'},
        title='Distribution by Ecological Level'
    )
    display(fig)
else:
    print("Ecological_level column not found")